include/merlin/group_lock.hpp

```diff
@@ -0,0 +1,213 @@
+      }
+      reader_count_.fetch_sub(1, std::memory_order_acq_rel);
+    }
+
+    /* Ban reader */
+    for (;;) {
+      while (reader_count_.load(std::memory_order_acquire) > 1) {
+      }
+      writer_count_.fetch_add(1, std::memory_order_acq_rel);
+      if (reader_count_.load(std::memory_order_acquire) == 1) {
+        break;
+      }
+      writer_count_.fetch_sub(1, std::memory_order_acq_rel);
+    }
+  }
+
+  void unlock_both() noexcept {
+    reader_count_.fetch_sub(1, std::memory_order_release);
+    writer_count_.fetch_sub(1, std::memory_order_release);
+    unique_flag_.store(false, std::memory_order_release);
+  }
+
+  int writer_count() noexcept {
+    return writer_count_.load(std::memory_order_relaxed);
+  }
+
+  int reader_count() noexcept {
+    return reader_count_.load(std::memory_order_relaxed);
+  }
+
+ private:
+  std::atomic<int> writer_count_;
+  std::atomic<int> reader_count_;
+  std::atomic<bool> unique_flag_;
+};
+
+class reader_shared_lock {
+ public:
+  reader_shared_lock(const reader_shared_lock&) = delete;
+  reader_shared_lock& operator=(const reader_shared_lock&) = delete;
+
+  explicit reader_shared_lock(group_shared_mutex& mutex) : mutex_(&mutex) {
+    mutex_->lock_read();
+    owns_ = true;
+  }
+
+  explicit reader_shared_lock(group_shared_mutex& mutex, std::defer_lock_t)
+      : mutex_(&mutex), owns_(false) {}
+
+  ~reader_shared_lock() {
+    if (owns_) {
+      mutex_->unlock_read();
+    }
+  }
+
+  void lock() noexcept {
+    if (!owns_) {
+      mutex_->lock_read();
+      owns_ = true;
+    }
+  }
+
+  bool owns_lock() const noexcept { return owns_; }
+
+ private:
+  group_shared_mutex* const mutex_;
+  bool owns_;
+};
+
+class writer_shared_lock {
+ public:
+  writer_shared_lock(const writer_shared_lock&) = delete;
+  writer_shared_lock& operator=(const writer_shared_lock&) = delete;
+
+  explicit writer_shared_lock(group_shared_mutex& mutex) : mutex_(&mutex) {
+    mutex_->lock_write();
+    owns_ = true;
+  }
+
+  explicit writer_shared_lock(group_shared_mutex& mutex, std::defer_lock_t)
+      : mutex_(&mutex), owns_(false) {}
+
+  ~writer_shared_lock() {
+    if (owns_) {
+      mutex_->unlock_write();
+    }
+  }
+
+  void lock() noexcept {
+    if (!owns_) {
+      mutex_->lock_write();
+      owns_ = true;
+    }
+  }
+
+  bool owns_lock() const noexcept { return owns_; }
+
+ private:
+  group_shared_mutex* const mutex_;
+  bool owns_;
+};
+
+class unique_lock {
+ public:
+  unique_lock(const unique_lock&) = delete;
+  unique_lock& operator=(const unique_lock&) = delete;
+
+  explicit unique_lock(group_shared_mutex& mutex) : mutex_(&mutex) {
+    mutex_->lock_both();
+    owns_ = true;
+  }
+
+  explicit unique_lock(group_shared_mutex& mutex, std::defer_lock_t) noexcept
+      : mutex_(&mutex), owns_(false) {}
+
+  ~unique_lock() {
+    if (owns_) {
+      mutex_->unlock_both();
+    }
+  }
+
+  void lock() {
+    if (owns_) {
+      throw(std::errc::resource_deadlock_would_occur);
```

Comment:
No throw, use an assertion instead.

Reply:
Per my knowledge, the assert is usually used for debug mode. We're trying to align the activity of group lock with the `std::shared_lock` and 'std::unique_lock', which throws the same exception. We can refer to https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/unique_lock.h#L138.

Reply:
Accept as we discuss offline.